lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs

```diff
@@ -91,14 +100,46 @@ public byte[] Call(byte[] memento)
              
                 _dataReducer.Send(result);
             }
+            _stopped = true;
             return null;
         }
 
         /// <summary>
+        /// Received close event. 
+        /// If it is closed event sent from driver, set _break to true so that to stop the iteration in Call method
+        /// After _enforceCloseTimeout, if the task is still not closed, throw IMRUTaskSystemException to enforce stop. 
+        /// </summary>
+        /// <param name="closeEvent"></param>
+        public void OnNext(ICloseEvent closeEvent)
+        {
+            if (closeEvent.Value.IsPresent() && Encoding.UTF8.GetString(closeEvent.Value.Value).Equals(TaskManager.CloseTaskByDriver))
+            {
+                Logger.Log(Level.Info, "The task received close event with message: {0}.", Encoding.UTF8.GetString(closeEvent.Value.Value));
+                _break = true;
+            }
+            Thread.Sleep(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
```

Comment:
there are only three reasons it will not come out:
a) GC does not throw exceptions when stream fails - this is limitation of current REEF and need to be addressed.
b) Map or Update function hangs inside - This is a bug in user code and we are not responsible for it and should not care for it.
c) Any other reason - This is either a bug in REEF or our Host tasks and is again a REEF problem.

If anything, these sleeps will hide any real bugs in future.